independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/MemoryClassPathElement.java

```diff
@@ -22,7 +23,21 @@ public MemoryClassPathElement(Map<String, byte[]> resources) {
     }
 
     public void reset(Map<String, byte[]> resources) {
-        this.resources = resources;
+        Map<String, byte[]> newResources = new HashMap<>(resources);
+        //we can't delete .class files from the loader
+        //gizmo may not generate the same function names on restart
+        //so if we delete them already loaded classes may have problems, as functions they reference
+        //may have been removed
+        //see https://github.com/quarkusio/quarkus/issues/8301
+        for (Map.Entry<String, byte[]> e : this.resources.entrySet()) {
+            if (newResources.containsKey(e.getKey())) {
+                continue;
+            }
+            if (e.getKey().endsWith(".class")) {
```

Comment:
I wonder if we should make the `io.quarkus.gizmo.BytecodeCreatorImpl.FUNCTION` field public and use it here?